syntax = "proto3";

package metatrader;

service PositionsService {

    // Gets an array of the current open positions.
    rpc GetPositions (GetPositionsReq) returns (PositionsList);

    // Opens a position
    rpc OpenPosition (OpenPositionReq) returns (OpenPositionRes);

    // Modifies a position
    rpc ModifyPosition (ModifyPositionReq) returns (ModifyPositionRes);

    // Closes a position
    rpc ClosePosition (ClosePositionReq) returns (ClosePositionRes);
}

enum PositionType {
  LONG = 0;
  SHORT = 1;
}

message Position {
    int64 id = 1;
    string symbol = 2;
    PositionType type = 3;
    float price = 4;
    float sl = 5;
    float tp = 6;
    float volume = 7;
    int64 time = 9;
    int64 magic = 11;
    string comment = 12;
}

message GetPositionsReq {
    optional string symbol = 1;
}

message PositionsList {
    repeated Position positions = 1;
}

message OpenPositionReq {
    PositionType type = 1;
    string symbol = 2;
    float volume = 3;
    float sl = 4;
    float tp = 5;
}

message OpenPositionRes {
    int64 ticket = 1;
}

message ModifyPositionReq {
    int64 ticket = 1;
    float sl = 2;
    float tp = 3;
}

message ModifyPositionRes {
    bool status = 1;
}

message ClosePositionReq {
    int64 ticket = 1;
}

message ClosePositionRes {
    bool status = 1;
}
